src/osd/PrimaryLogPG.cc

```diff
@@ -3991,15 +3991,16 @@ int PrimaryLogPG::do_extent_cmp(OpContext *ctx, OSDOp& osd_op)
     return result;
   }
 
-  if (read_op.outdata.length() != osd_op.indata.length())
-    return -EINVAL;
-
-  for (uint64_t p = 0; p < osd_op.indata.length(); p++) {
+  for (uint64_t p = 0; p < read_op.outdata.length(); p++) {
     if (read_op.outdata[p] != osd_op.indata[p]) {
       return (-MAX_ERRNO - p);
     }
   }
 
+  if (read_op.outdata.length() != osd_op.indata.length()) {
```

Comment:
Thinking about this more, even though I think this is likely the correct behavior for the OSD (i.e. you said to compare beyond the extent of the object), this will complicate RBD where truncated/missing objects are treated as zeroed. If a client app reads extents 0~512 of a non-existent object, sees that it's currently zero, and uses that zero buffer as a compare_and_write, the compare and write will fail with mismatch at offset 0. Therefore, I think we should keep the original loop bounds (on the ```indata```, but if ```p >= read_op.outdata.length()```, fail if ```indata[p] != 0```.